Adding docs for migrating from SDKv2 to the Framework (#461)
hashicorp/terraform-plugin-framework: website/docs/plugin/framework/migrating/attributes-blocks/validators-custom.mdx

---
page_title: 'Attribute Custom Validators: Migrating from SDKv2 to the Framework'
description: >-
  Validations check for required syntax, types, and acceptable values. Migrate custom attribute validation functions from SDKv2 to attribute validators in the Framework.
---

# Custom Validators

You can write custom validations that give users feedback about required syntax, types, and acceptable values in your
provider. The Framework has a collection of
[predefined validators](https://github.com/hashicorp/terraform-plugin-framework-validators). Refer to
[Predefined Validators](/plugin/framework/migrating/validators-predefined) to learn how to use them.

This page explains how to migrate attribute validation functions from SDKv2 to attribute validators in the Framework.

## SDKv2

In SDKv2, arbitrary validation logic can be applied to individual attributes by using `ValidateFunc` and/or
`ValidateDiagFunc`.

The following example shows the implementation of a validation that ensures that an integer attribute has a value
greater than one.

```go
func resourceExample() *schema.Resource {
    return &schema.Resource{
        /* ... */

>         Schema: map[string]*schema.Schema
>         }{

Reviewer: ```suggestion
        Schema: map[string]*schema.Schema{
```

I think?

            "attribute_example": {
                ValidateDiagFunc: validation.ToDiagFunc(validation.IntAtLeast(1)),
                /* ... */
```

## Framework

In the Framework, you implement either type of validation by setting the `Validators` field on the `tfsdk.Attribute`
struct with types that satisfy the `tfsdk.AttributeValidator` interface.

> The following example shows the implementation of a validation that ensures that an integer attribute has a value

Reviewer 2: ```suggestion
The following example shows how to implement a validation that ensures that an integer attribute has a value
```
